File: Algorithm_Rust/examples/3217. 从链表中移除在数组中存在的节点.rs
use std::collections::HashMap;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl Solution {
    pub fn modified_list(nums: Vec<i32>, head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut map = HashMap::new();
        nums.iter().for_each(|el| {
            map.insert(*el, true);
        });
        let mut head = head;
        let mut new_head = Box::new(ListNode::new(0));
        let mut node = &mut new_head;
        while let Some(mut h) = head {
            head = h.next.take();
            if !map.contains_key(&h.val) {
                node.next = Some(h);
                node = node.next.as_mut().unwrap();
            }
        }
        new_head.next
    }
}

fn main() {}

File: Algorithm_Rust/examples/3217. 从链表中移除在数组中存在的节点.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function modifiedList(nums: number[], head: ListNode | null): ListNode | null {
  let map = new Map<number, boolean>();
  nums.forEach((el) => map.set(el, true));
  let new_head = new ListNode();
  let node = new_head;
  while (head) {
    if (!map.get(head.val)) {
      node.next = head;
      node = node.next;
    }
    head = head.next;
  }
  node.next = null;
  return new_head.next;
}

export {};
